### PR TITLE
Pass `TyCtxt` to early diagostics decoration

### PR DESCRIPTION
**Pull request by Urgau**
This PR pass a `TyCtxt` to the early diagnostics decoration code so that diagnostics code that take advantage of (a very limited but still useful) `TyCtxt` in their note, help, suggestions, ...

This is particulary useful for #133221 which wants to get the crate name of a `DefId`, which is possible with `tcx.crate_name(...)`.

I highly recommend reviewing this PR commit by commit.

r? @jieyouxu

### CODE REVIEW

File: compiler/rustc_lint/src/early.rs [hunk 1]
```diff
@@ -8,36 +8,73 @@ use rustc_ast::ptr::P;
...
 }) }
 
 /// Implements the AST traversal for early lint passes. `T` provides the
 /// `check_*` methods.
-pub struct EarlyContextAndPass<'a, T: EarlyLintPass> {
+pub struct EarlyContextAndPass<'a, 'b, T: EarlyLintPass> {
     context: EarlyContext<'a>,
+    tcx: Option<TyCtxt<'b>>,
     pass: T,
 }
```
Comment (jieyouxu) [inline compiler/rustc_lint/src/early.rs:34]: Suggestion: can you use `'ecx` and `'tcx` here?

File: compiler/rustc_lint/src/early.rs [hunk 2]
```diff
@@ -67,7 +104,7 @@ impl<'a, T: EarlyLintPass> EarlyContextAndPass<'a, T> {
     }
 }
 
-impl<'a, T: EarlyLintPass> ast_visit::Visitor<'a> for EarlyContextAndPass<'a, T> {
+impl<'a, 'b, T: EarlyLintPass> ast_visit::Visitor<'a> for EarlyContextAndPass<'a, 'b, T> {
```
Comment (jieyouxu) [inline compiler/rustc_lint/src/early.rs:75]: Suggestion: ditto on the lifetime names

File: compiler/rustc_lint/src/early.rs [hunk 3]
```diff
@@ -313,7 +350,7 @@ pub trait EarlyCheckNode<'a>: Copy {
     fn attrs<'b>(self) -> &'b [ast::Attribute]
     where
         'a: 'b;
-    fn check<'b, T: EarlyLintPass>(self, cx: &mut EarlyContextAndPass<'b, T>)
+    fn check<'b, 'c, T: EarlyLintPass>(self, cx: &mut EarlyContextAndPass<'b, 'c, T>)
```
Comment (jieyouxu) [inline compiler/rustc_lint/src/early.rs:321]: Question: sorry, where did `'c` come from...? Or is this `'tcx`?

**Review by jieyouxu — CHANGES_REQUESTED** [review]
Thanks, I think this seems reasonable. I have some suggestions regarding the lifetime names (random `'a`, `'b` and `'c` are really confusing).

File: compiler/rustc_lint/src/early.rs (hunk 3 above)
Comment (Urgau) [reply compiler/rustc_lint/src/early.rs]: It's `'tcx` ;-)

**Comment by Urgau** [comment]
I agree the lifetimes are confusing, so as suggested I cleaned them up (including some bounds). Pushed that cleanup in https://github.com/rust-lang/rust/pull/134339/commits/e3e5bd95cd306160cfa799b3717583fec01c7793.

@rustbot ready

**Review by jieyouxu — APPROVED** [review]
Thanks, r=me once PR CI is green.

**Pull request merged**